File: assets/js/delta/scene.js
// scene.js — work cell (table, parts, bins, conveyor). Mirror of python/deltarobot/scene.py.
import { workCylinder } from './kinematics.js';

export const GRAB_Z_TOL = 0.012;
const r4 = (v) => Math.round(v * 1e4) / 1e4;

export const SCENE_KINDS = { empty: '빈 작업대', pick_place: '분류 (pick & place)', conveyor: '컨베이어 트래킹', drawing: '그리기 (펜)' };

export function defaultScene(d, kind = 'pick_place', cyl = null) {
  const height = 0.25 * (d.upper_arm + d.forearm);
  cyl = cyl || workCylinder(d, height);
  const rc = Math.max(0.02, cyl.diameter / 2);
  const surface = cyl.z_bottom - d.toolLength;
  const size = Math.min(0.06, Math.max(0.015, 0.12 * rc));
  const h = 0.6 * size;
  const scene = { kind, surface_z: r4(surface), work_radius: r4(rc), work_top: r4(cyl.z_top - d.toolLength),
    parts: [], bins: [], conveyor: null };
  const colors = ['red', 'blue', 'green'];
  if (kind === 'pick_place') {
    let k = 0;
    for (let row = 0; row < 2; row++) {
      for (let col = 0; col < 3; col++) {
        scene.parts.push({ id: 'p' + (k + 1), x: r4(-0.55 * rc + row * 0.22 * rc), y: r4((col - 1) * 0.3 * rc),
          size: r4(size), h: r4(h), color: colors[col], material: row === 1 ? 'steel' : 'plastic' });
        k++;
      }
    }
    colors.forEach((c, j) => scene.bins.push({ id: 'ABC'[j], x: r4(0.5 * rc), y: r4((j - 1) * 0.36 * rc),
      w: r4(0.3 * rc), d: r4(0.3 * rc), h: r4(0.8 * h), color: c }));
  } else if (kind === 'conveyor') {
    const offs = [0, 0.12, -0.12, 0.06, -0.06];
    const spawn = [];
    for (let n = 0; n < 30; n++) {
      spawn.push({ t: r4(1 + 2 * n), y_off: r4(offs[n % 5] * rc), color: colors[n % 2], material: 'plastic' });
    }
    scene.conveyor = { y: r4(-0.45 * rc), width: r4(0.45 * rc), x_start: r4(-1.3 * rc), x_end: r4(1.3 * rc),
      speed: r4(0.2 * rc), size: r4(size), h: r4(h), spawn };
    colors.slice(0, 2).forEach((c, j) => scene.bins.push({ id: 'AB'[j], x: r4((j - 0.5) * 0.7 * rc), y: r4(0.5 * rc),
      w: r4(0.35 * rc), d: r4(0.3 * rc), h: r4(0.8 * h), color: c }));
  } else if (kind === 'drawing') {
    scene.paper = { x: 0, y: 0, w: r4(1.4 * rc), d: r4(1.4 * rc) };
  }
  return scene;
}

export class SceneState {
  constructor(data) {
    this.data = data ? JSON.parse(JSON.stringify(data)) : { kind: 'empty', surface_z: -1, parts: [], bins: [], conveyor: null };
    this.surface_z = Number(this.data.surface_z ?? -1);
    this.parts = (this.data.parts || []).map((p) => ({ ...p, mode: 'table', z: this.surface_z }));
    const conv = this.data.conveyor;
    if (conv) {
      (conv.spawn || []).forEach((s, n) => this.parts.push({
        id: 'c' + (n + 1), size: conv.size, h: conv.h, color: s.color || 'red', material: s.material || 'plastic',
        mode: 'conveyor', x0: conv.x_start, t0: s.t, y: conv.y + (s.y_off || 0), z: this.surface_z,
      }));
    }
    this.held = null;
    this.events = [];
  }

  pos(p, t) {
    if (p.mode === 'gone' || p.mode === 'held') return null;
    if (p.mode === 'conveyor') {
      if (t < p.t0) return null;
      const conv = this.data.conveyor;
      const x = p.x0 + conv.speed * (t - p.t0);
      if (x > conv.x_end) return null;
      return [x, p.y, p.z];
    }
    return [p.x, p.y, p.z];
  }

  visible(t) {
    const out = [];
    for (const p of this.parts) {
      const q = this.pos(p, t);
      if (q) out.push({ part: p, x: q[0], y: q[1], z: q[2] });
    }
    return out;
  }

  grab(tcp, t, tool) {
    if (this.held || tool.kind === 'pen') return null;
    let best = null, bestD = 1e9;
    for (const p of this.parts) {
      const q = this.pos(p, t);
      if (!q) continue;
      if (tool.ferrous_only && p.material !== 'steel') continue;
      const dxy = Math.hypot(tcp[0] - q[0], tcp[1] - q[1]);
      const top = q[2] + p.h;
      const zok = tool.kind === 'gripper'
        ? (q[2] - 0.002 <= tcp[2] && tcp[2] <= top + GRAB_Z_TOL)
        : Math.abs(tcp[2] - top) <= GRAB_Z_TOL;
      if (zok && dxy <= Math.max(0.012, 0.45 * p.size) && dxy < bestD) { best = p; bestD = dxy; }
    }
    if (!best) { this.events.push({ t, type: 'grab_miss' }); return null; }
    best.mode = 'held';
    this.held = best;
    this.events.push({ t, type: 'grab', id: best.id });
    return best.id;
  }

  release(tcp, t) {
    const p = this.held;
    if (!p) return null;
    this.held = null;
    const x = tcp[0], y = tcp[1];
    p.x = x; p.y = y;
    for (const b of this.data.bins || []) {
      if (Math.abs(x - b.x) <= b.w / 2 && Math.abs(y - b.y) <= b.d / 2) {
        Object.assign(p, { mode: 'bin', z: this.surface_z + 0.003, bin: b.id, bin_color: b.color });
        this.events.push({ t, type: 'place', id: p.id, bin: b.id });
        return p.id;
      }
    }
    const conv = this.data.conveyor;
    if (conv && Math.abs(y - conv.y) <= conv.width / 2 && x >= conv.x_start && x <= conv.x_end) {
      Object.assign(p, { mode: 'conveyor', x0: x, t0: t, y, z: this.surface_z });
    } else {
      Object.assign(p, { mode: 'table', z: this.surface_z });
    }
    this.events.push({ t, type: 'drop', id: p.id });
    return p.id;
  }

  score() {
    const per = {};
    for (const b of this.data.bins || []) per[b.id] = 0;
    let correct = 0, total = 0;
    for (const p of this.parts) {
      if (p.mode === 'bin') {
        per[p.bin] = (per[p.bin] || 0) + 1;
        total++;
        if (p.bin_color === p.color) correct++;
      }
    }
    return { per_bin: per, correct, total };
  }
}

File: python/deltarobot/scene.py
"""Work cell scene: table, parts, bins and an optional conveyor.

The scene is plain JSON so the web simulator, the Playground and PC Python share it:

    {"kind": "pick_place", "surface_z": -0.33,
     "parts": [{"id": "p1", "x": .., "y": .., "size": 0.03, "h": 0.018,
                "color": "red", "material": "plastic"}],
     "bins": [{"id": "A", "x": .., "y": .., "w": .., "d": .., "h": .., "color": "red"}],
     "conveyor": {"y": .., "width": .., "x_start": .., "x_end": .., "speed": ..,
                  "spawn": [{"t": 0.0, "y_off": 0.0, "color": "red", "material": "plastic"}, ...]}
                 | null}

Grab / release rules are mirrored in assets/js/delta/scene.js.
"""

from __future__ import annotations

import copy
import math
from typing import Any, Dict, List, Optional, Sequence

from . import kinematics as kin

GRAB_Z_TOL = 0.012
PART_SIZE = 0.03


def _round(v: float) -> float:
    """Same as JS Math.round(v * 1e4) / 1e4 (half up, not banker's rounding)."""
    return math.floor(v * 1e4 + 0.5) / 1e4


def default_scene(design, kind: str = "pick_place") -> Dict[str, Any]:
    """Scene sized to the design's work cylinder (same numbers as the web simulator)."""
    height = 0.25 * (design.upper_arm + design.forearm)
    cyl = kin.work_cylinder(design, height)
    rc = max(0.02, cyl["diameter"] / 2.0)
    surface = cyl["z_bottom"] - design.tool_length
    size = min(0.06, max(0.015, 0.12 * rc))
    h = 0.6 * size
    scene: Dict[str, Any] = {"kind": kind, "surface_z": _round(surface), "work_radius": _round(rc),
                             "work_top": _round(cyl["z_top"] - design.tool_length),
                             "parts": [], "bins": [], "conveyor": None}
    colors = ["red", "blue", "green"]
    if kind == "pick_place":
        k = 0
        for row in range(2):
            for col in range(3):
                scene["parts"].append({
                    "id": "p%d" % (k + 1),
                    "x": _round(-0.55 * rc + row * 0.22 * rc),
                    "y": _round((col - 1) * 0.3 * rc),
                    "size": _round(size), "h": _round(h),
                    "color": colors[col], "material": "steel" if row == 1 else "plastic"})
                k += 1
        for j, c in enumerate(colors):
            scene["bins"].append({"id": "ABC"[j], "x": _round(0.5 * rc), "y": _round((j - 1) * 0.36 * rc),
                                  "w": _round(0.3 * rc), "d": _round(0.3 * rc), "h": _round(0.8 * h),
                                  "color": c})
    elif kind == "conveyor":
        spawn = []
        offs = [0.0, 0.12, -0.12, 0.06, -0.06]
        for n in range(30):
            spawn.append({"t": _round(1.0 + 2.0 * n), "y_off": _round(offs[n % 5] * rc),
                          "color": colors[n % 2], "material": "plastic"})
        scene["conveyor"] = {"y": _round(-0.45 * rc), "width": _round(0.45 * rc),
                             "x_start": _round(-1.3 * rc), "x_end": _round(1.3 * rc),
                             "speed": _round(0.2 * rc), "size": _round(size), "h": _round(h),
                             "spawn": spawn}
        for j, c in enumerate(colors[:2]):
            scene["bins"].append({"id": "AB"[j], "x": _round((j - 0.5) * 0.7 * rc), "y": _round(0.5 * rc),
                                  "w": _round(0.35 * rc), "d": _round(0.3 * rc), "h": _round(0.8 * h),
                                  "color": c})
    elif kind == "drawing":
        scene["paper"] = {"x": 0.0, "y": 0.0, "w": _round(1.4 * rc), "d": _round(1.4 * rc)}
    return scene


class Scene:
    """Mutable scene state driven by tool on/off events."""

    def __init__(self, data: Optional[Dict[str, Any]] = None):
        self.data = copy.deepcopy(data) if data else {"kind": "empty", "surface_z": -1.0,
                                                       "parts": [], "bins": [], "conveyor": None}
        self.surface_z = float(self.data.get("surface_z", -1.0))
        self._parts: List[Dict[str, Any]] = []
        for p in self.data.get("parts", []):
            q = dict(p)
            q.update(mode="table", z=self.surface_z)
            self._parts.append(q)
        conv = self.data.get("conveyor")
        if conv:
            for n, s in enumerate(conv.get("spawn", [])):
                self._parts.append({"id": "c%d" % (n + 1), "size": conv["size"], "h": conv["h"],
                                    "color": s.get("color", "red"), "material": s.get("material", "plastic"),
                                    "mode": "conveyor", "x0": conv["x_start"], "t0": s["t"],
                                    "y": conv["y"] + s.get("y_off", 0.0), "z": self.surface_z})
        self.held: Optional[Dict[str, Any]] = None
        self.events: List[Dict[str, Any]] = []

    # ------------------------------------------------------------ queries
    def _pos(self, p: Dict[str, Any], t: float) -> Optional[List[float]]:
        mode = p["mode"]
        if mode == "gone" or mode == "held":
            return None
        if mode == "conveyor":
            if t < p["t0"]:
                return None
            conv = self.data["conveyor"]
            x = p["x0"] + conv["speed"] * (t - p["t0"])
            if x > conv["x_end"]:
                return None
            return [x, p["y"], p["z"]]
        return [p["x"], p["y"], p["z"]]

    def parts(self, t: float = 0.0) -> List[Dict[str, Any]]:
        """Visible (not held) parts at time t: id, x, y, z (bottom), top, size, color, material, on."""
        out = []
        for p in self._parts:
            pos = self._pos(p, t)
            if pos is None:
                continue
            out.append({"id": p["id"], "x": pos[0], "y": pos[1], "z": pos[2], "top": pos[2] + p["h"],
                        "size": p["size"], "h": p["h"], "color": p["color"], "material": p["material"],
                        "on": p["mode"]})
        return out

    def bins(self) -> List[Dict[str, Any]]:
        return [dict(b) for b in self.data.get("bins", [])]

    def bin(self, bin_id: str) -> Dict[str, Any]:
        for b in self.data.get("bins", []):
            if b["id"] == bin_id:
                return dict(b)
        raise KeyError("no bin %r" % bin_id)

    def conveyor(self) -> Optional[Dict[str, Any]]:
        c = self.data.get("conveyor")
        return dict(c) if c else None

    def score(self) -> Dict[str, Any]:
        """Parts per bin and how many ended in the bin of the same colour."""
        per: Dict[str, int] = {b["id"]: 0 for b in self.data.get("bins", [])}
        correct = 0
        for p in self._parts:
            if p["mode"] == "bin":
                per[p["bin"]] = per.get(p["bin"], 0) + 1
                if p.get("bin_color") == p["color"]:
                    correct += 1
        return {"per_bin": per, "correct": correct, "total": sum(per.values())}

    # ------------------------------------------------------------ tool events
    def grab(self, tcp: Sequence[float], t: float, tool: Dict[str, Any]) -> Optional[str]:
        """Tool switched on at TCP position; returns the id of the attached part."""
        if self.held is not None or tool.get("kind") == "pen":
            return None
        best, best_d = None, 1e9
        for p in self._parts:
            pos = self._pos(p, t)
            if pos is None:
                continue
            if tool.get("ferrous_only") and p["material"] != "steel":
                continue
            dxy = math.hypot(tcp[0] - pos[0], tcp[1] - pos[1])
            top = pos[2] + p["h"]
            if tool.get("kind") == "gripper":
                zok = pos[2] - 0.002 <= tcp[2] <= top + GRAB_Z_TOL
            else:
                zok = abs(tcp[2] - top) <= GRAB_Z_TOL
            if zok and dxy <= max(0.012, 0.45 * p["size"]) and dxy < best_d:
                best, best_d = p, dxy
        if best is None:
            self.events.append({"t": t, "type": "grab_miss"})
            return None
        best["mode"] = "held"
        self.held = best
        self.events.append({"t": t, "type": "grab", "id": best["id"]})
        return best["id"]

    def release(self, tcp: Sequence[float], t: float) -> Optional[str]:
        p = self.held
        if p is None:
            return None
        self.held = None
        x, y = tcp[0], tcp[1]
        p["x"], p["y"] = x, y
        for b in self.data.get("bins", []):
            if abs(x - b["x"]) <= b["w"] / 2 and abs(y - b["y"]) <= b["d"] / 2:
                p.update(mode="bin", z=self.surface_z + 0.003, bin=b["id"], bin_color=b.get("color"))
                self.events.append({"t": t, "type": "place", "id": p["id"], "bin": b["id"]})
                return p["id"]
        conv = self.data.get("conveyor")
        if conv and abs(y - conv["y"]) <= conv["width"] / 2 and conv["x_start"] <= x <= conv["x_end"]:
            p.update(mode="conveyor", x0=x, t0=t, y=y, z=self.surface_z)
        else:
            p.update(mode="table", z=self.surface_z)
        self.events.append({"t": t, "type": "drop", "id": p["id"]})
        return p["id"]
